registry/storage/driver/s3-aws/s3.go

```diff
@@ -942,15 +942,152 @@ func (d *driver) Walk(ctx context.Context, from string, f storagedriver.WalkFn)
+		// Some s3 implementations (looking at you ceph/rgw) have a buggy
+		// implementation so we intionally avoid ever using it, preferring instead
+		// to calculate the count from the Contents and CommonPrefixes fields of
+		// the s3.ListObjectsV2Output. we retain the commented out KeyCount code
+		// and this comment so as not to forget this problem moving forward.
+		//
+		// count = *objects.KeyCount
+		// *objectCount += *objects.KeyCount
+		count := int64(len(objects.Contents) + len(objects.CommonPrefixes))
+		*objectCount += count
+
+		walkInfos := make([]walkInfoContainer, 0, count)
+
+		for _, dir := range objects.CommonPrefixes {
+			commonPrefix := *dir.Prefix
+			walkInfos = append(walkInfos, walkInfoContainer{
+				prefix: dir.Prefix,
+				FileInfoFields: storagedriver.FileInfoFields{
+					IsDir: true,
+					Path:  strings.Replace(commonPrefix[:len(commonPrefix)-1], d.s3Path(""), prefix, 1),
+				},
+			})
+		}
+
+		for _, file := range objects.Contents {
+			// empty prefixes are listed as objects inside its own prefix.
+			// https://docs.aws.amazon.com/AmazonS3/latest/user-guide/using-folders.html
+			if strings.HasSuffix(*file.Key, "/") {
+				continue
+			}
+			walkInfos = append(walkInfos, walkInfoContainer{
+				FileInfoFields: storagedriver.FileInfoFields{
+					IsDir:   false,
+					Size:    *file.Size,
+					ModTime: *file.LastModified,
+					Path:    strings.Replace(*file.Key, d.s3Path(""), prefix, 1),
+				},
+			})
+		}
+
+		sort.SliceStable(walkInfos, func(i, j int) bool { return walkInfos[i].FileInfoFields.Path < walkInfos[j].FileInfoFields.Path })
+
+		for _, walkInfo := range walkInfos {
+			err := f(walkInfo)
+
+			if err == storagedriver.ErrSkipDir {
+				if walkInfo.IsDir() {
+					continue
+				} else {
+					// is file, stop gracefully
+					retStop = true
+					return false
+				}
+			} else if err != nil {
+				retError = err
+				return false
+			}
+
+			if walkInfo.IsDir() {
+				if stop, err := d.doWalkRecursive(ctx, objectCount, *walkInfo.prefix, prefix, f); err != nil || stop {
+					retError = err
+					retStop = stop
+					return false
+				}
+			}
+		}
+		return true
+	})
+
+	if retError != nil {
+		return retStop, retError
+	}
+
+	if listObjectErr != nil {
+		return retStop, listObjectErr
+	}
+
+	return retStop, nil
+}
+
+// doWalkWithoutSkipDir is an implementation of Walk that is optimized for listing objects and suboptimal for contexts
+// where storagedriver.ErrSkipDir is utilized to skip over large numbers of keys.
+func (d *driver) doWalkWithoutSkipDir(parentCtx context.Context, objectCount *int64, path, prefix string, f storagedriver.WalkFn) error {
 	var (
 		retError error
 		// the most recent directory walked for de-duping
 		prevDir string
 		// the most recent skip directory to avoid walking over undesirable files
 		prevSkipDir string
 	)
-	prevDir = prefix + path
+
+	prevDir = strings.Replace(prefix+path, d.s3Path(""), prefix, 1)
```

Comment:
this change prevents listing the relative `path`

Reply:
Could you put this comment in a code comment with a short example demonstrating the cases it is intended to prevent?